text/3729-sized-hierarchy.md

```diff
@@ -0,0 +1,1412 @@
+have SIMD registers of a known, fixed length and a variety of intrinsics
+which operate on these registers. For example, x86-64 introduced 128-bit SIMD
+registers with SSE, 256-bit SIMD registers with AVX, and 512-bit SIMD registers
+with AVX-512, and Arm introduced 128-bit SIMD registers with Neon.
+
+As an alternative to releasing SIMD extensions with greater bit widths, Arm and
+RISC-V have vector extensions (SVE/Scalable Vector Extension and the "V" Vector
+Extension/RVV respectively) where the bit width of vector registers depends on the
+CPU implementation, and the instructions which operate these registers are bit
+width-agnostic.
+
+As a consequence, these types are not `Sized` in the Rust sense, as the size of
+a scalable vector cannot be known during compilation, but is a runtime constant.
+For example, the size of these types could be determined by inspecting the value
+in a register - this is not available at compilation time and the value may
+differ between any given CPU implementation. Both SVE and RVV have mechanisms to
+change the system's vector length (up to the maximum supported by the CPU
+implementations) but this is not supported by the proposed ABI for these types.
+
+However, despite not implementing `Sized`, these are value types which should
+implement `Copy` and can be returned from functions, can be variables on the
+stack, etc. These types should implement `Copy` but given that `Sized` is a
+supertrait of `Copy`, they cannot be `Copy` without being `Sized`, and
+they aren't `Sized`.
+
+Introducing a `const Sized` trait will enable `Copy` to be implemented for
+those types whose size is a runtime constant to function correctly without
+special cases in the type system. See
+[rfcs#3268: Scalable Vectors][rfc_scalable_vectors].
+
+## Unsized types and extern types
+[unsized-types-and-extern-types]: #unsized-types-and-extern-types
+
+[Extern types][rfc_extern_types] has long been blocked on these types being
+neither `Sized` nor `?Sized` ([relevant issue][issue_extern_types_align_size]).
+
+RFC #1861 defined that `std::mem::size_of_val` and `std::mem::align_of_val`
+should not be defined for extern types but not how this should be achieved, and
+suggested an initial implementation could panic when called with extern types,
+but this is always wrong, and not desirable behavior in keeping with Rust's
+values. `size_of_val` returns 0 and `align_of_val` returns 1 for extern types in
+the current implementation. Ideally `size_of_val` and `align_of_val` would error
+if called with an extern type, but this cannot be expressed in the bounds of
+`size_of_val` and `align_of_val` and this remains a blocker for extern types.
+
+Furthermore, unsized types cannot be members of structs as their alignment is
+unknown and this is necessary to calculate field offsets. Extern types also
+cannot be used in `Box` as `Box` requires size and alignment for both allocation
+and deallocation.
```

Comment:
We currently allow extern types to appear as the first field of a struct. There's some unstable code relying on that, even in the compiler itself.

Reply:
Fixed!

Reply:
I think you meant last field though, that's what it seemed like was true from some quick testing, and that's what I've written. Happy to change it if I'm mistaken.

Reply:
It also has to be the last field as usual for unsized types. But that's not enough, since as you said in general, computing the offset requires knowing the alignment.

The current entirely ad-hoc solution to this is to panic during the offset computation of the offset is not `0`. If it is `0`, then the alignment does not matter for offset computation.